### PR TITLE
Andrey Rohau / Refactor email notification

### PR DESCRIPTION
**Pull request by andreyrogov8**

### CODE REVIEW

File: Elevel/Application/Infrastructure/Configurations/EmailFormConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Elevel.Application.Infrastructure.Configurations
+{
+    public class EmailFormConfiguration
+    {
+        public List<MailboxAddress> ReceiverEmail { get; set; } = new List<MailboxAddress>();
```
Comment (parys) [inline Elevel/Application/Infrastructure/Configurations/EmailFormConfiguration.cs:12]: please plural receiversEmails

File: Elevel/Elevel.Infrastructure.Services/Implementation/MailService.cs [hunk 2]
```diff
@@ -109,25 +58,25 @@ public string SendMessage(Guid receiverId, string subject, string body)
...
+                }
+
+                _message.To.AddRange(email.ReceiverEmail);
+                _message.Subject = email.Subject;
+                _message.Body = new TextPart("plain")
+                {
+                    Text = email.Body
+                };
 
                 try
```
Comment (parys) [inline Elevel/Elevel.Infrastructure.Services/Implementation/MailService.cs:81]: empty code?

File: Elevel/Elevel.Infrastructure.Services/Jobs/EmailJob.cs [hunk 3]
```diff
@@ -16,21 +19,81 @@ namespace Elevel.Infrastructure.Services.Jobs
...
         }
 
         public async Task Execute(IJobExecutionContext context)
         {
-            await _mail.SendNotificationsToHrsAndUsersAsync();
+            var userEmails = await _context.Tests
+                .Where(x => x.AssignmentEndDate.HasValue
+                    && DateTimeOffset.Compare(x.AssignmentEndDate.Value.Date, DateTimeOffset.UtcNow.Date) >= 0)
+                .AsNoTracking()
+                .Select(x => MailboxAddress.Parse(
```
Comment (parys) [inline Elevel/Elevel.Infrastructure.Services/Jobs/EmailJob.cs:42]: why not include ?

File: Elevel/Elevel.Infrastructure.Services/Jobs/EmailJob.cs [hunk 4]
```diff
@@ -16,21 +19,81 @@ namespace Elevel.Infrastructure.Services.Jobs
...
+                .ToListAsync();
+
+            var missedDeadlineUsers = _context.Tests
+                .Where(x => x.AssignmentEndDate.HasValue
+                    && DateTimeOffset.Compare(x.AssignmentEndDate.Value.Date, DateTimeOffset.UtcNow.Date) < 0
+                    && !x.GrammarMark.HasValue)
+                .Select(x => new
+                {
+                    HrEmail = MailboxAddress.Parse(
+                    _userManager.Users.FirstOrDefault(u => u.Id == x.HrId).Email),
```
Comment (parys) [inline Elevel/Elevel.Infrastructure.Services/Jobs/EmailJob.cs:54]: 54 + 55 will generate 2 separate sql query per user
why not include ?

File: Elevel/Elevel.Infrastructure.Services/Jobs/EmailJob.cs [hunk 5]
```diff
@@ -16,21 +19,81 @@ namespace Elevel.Infrastructure.Services.Jobs
...
+
+            var userEmailForm = new EmailFormConfiguration();
+            if (userEmails.Count > 0)
+            {
+                userEmailForm.ReceiverEmail.AddRange(userEmails);
+                userEmailForm.Subject = "Elevel test reminder";
+                userEmailForm.Body = "You have assigned for test. \nCheck your profile.";
+                emailForms.Add(userEmailForm);
+            }
+            if (missedDict.Count > 0) {
```
Comment (parys) [inline Elevel/Elevel.Infrastructure.Services/Jobs/EmailJob.cs:84]: you can omit check, foreach will not execute for 0 items collection

File: Elevel/Elevel.Infrastructure.Services/Jobs/EmailScheduler.cs [hunk 6]
```diff
@@ -19,16 +19,16 @@ public static async Task Start(IServiceProvider service, SchedulerConfigurations
...
             ITrigger trigger = TriggerBuilder.Create()
                 .WithIdentity("MailingTrigger", "default")
-                .StartAt(DateBuilder.DateOf(
-                    schedulerConfig.startHour,
-                    schedulerConfig.startMinute, 
-                    schedulerConfig.startSecond,
-                    schedulerConfig.startDay,
-                    schedulerConfig.startMonth))
-                //.StartNow()
+                //.StartAt(DateBuilder.DateOf(
```
Comment (parys) [inline Elevel/Elevel.Infrastructure.Services/Jobs/EmailScheduler.cs]: why?

**Review by parys — APPROVED** [review]

**Review by 4500zenja1 — APPROVED** [review]

**Review by Evgenij-Kovpak — APPROVED** [review]

**Pull request merged**